iina/FFmpegController.swift

```diff
@@ -0,0 +1,80 @@
+//
+//  FFmpegController.swift
+//  iina
+//
+//  Created by Saagar Jha on 5/28/18.
+//  Copyright © 2018 lhc. All rights reserved.
+//
+
+import CoreGraphics
+
+struct FFmpegThumbnail {
+  let image: NSImage
+  let timestamp: Double
+}
+
+protocol FFmpegControllerDelegate {
+  func didGenerateThumbnails(_ thumbnails: [FFmpegThumbnail], forFileAtPath path: String, succeeded: Bool)
+  func didUpdateThumbnails(_ thumbnails: [FFmpegThumbnail]?, forFileAtPath path: String, withProgress progress: Int)
+}
+
+@objc(FFmpegController)
+class FFmpegController: NSObject {
+
+  private static let thumbnailQueue = DispatchQueue(label: "com.colliderli.iina.FFMpegController")
+  private var currentWorkItem: DispatchWorkItem!
+  private var thumbnails = [FFmpegThumbnail]()
+  private var timestamps = Set<Int64>()
+  private var lastDelegateCallbackTime = 0 as CFTimeInterval
+
+  public var delegate: FFmpegControllerDelegate? = nil
+  public var thumbnailCount: Int {
+    return thumbnails.count
+  }
+
+  public func generateThumbnailsForFile(atPath path: String) {
```

Comment:
I think the issue is the thumbnails array and timestamps set need to be emptied here.

Reply:
Oh, right, forgot to copy that over. Thanks for spotting that; I've updated the code.